frontend/controller/actions/identity.js

```diff
@@ -4,14 +4,33 @@ import sbp from '@sbp/sbp'
 import { GIErrorUIRuntimeError } from '@model/errors.js'
 import L, { LError } from '@view-utils/translations.js'
 import { imageUpload } from '@utils/image.js'
+import { pickWhere, difference } from '@utils/giLodash.js'
+import { captureLogsStart, captureLogsPause } from '~/frontend/model/captureLogs.js'
+import { SETTING_CURRENT_USER } from '~/frontend/model/database.js'
+import { LOGIN, LOGOUT } from '~/frontend/utils/events.js'
 import './mailbox.js'
 
 import { encryptedAction } from './utils.js'
 
+function generatedLoginState () {
+  const { contracts } = sbp('state/vuex/state')
+  return {
+    groupIds: Object.keys(pickWhere(contracts, ({ type }) => {
+      return type === 'gi.contracts/group'
+    }))
+  }
+}
+
+function diffLoginStates (s1: ?Object, s2: ?Object) {
+  if (typeof s1 !== 'object' || typeof s2 !== 'object') return true
+  const [g1, g2] = [(s1: Object).groupIds, (s2: Object).groupIds]
+  if (!g1 || !g2) return true
+  return (g1.length > g2.length ? difference(g1, g2) : difference(g2, g1)).length > 0
```

Comment:
Unless `.groupIds` might contain duplicate entries, checking whether `g1.length !== g2.length ` should be sufficient

Reply:
It's possible the lengths are the same but the groups are different

Reply:
Good point.